From 12249248c3d66b0588add9b168602544e9182698 Mon Sep 17 00:00:00 2001
From: Oliver Scott <olivercscott@gmail.com>
Date: Mon, 19 Apr 2021 14:48:01 -0400
Subject: [PATCH] Add setting to allow/disallow interacting with quick settings
 on secure lockscreen (1/2)

Do not handle tile clicks if keyguard is locked and setting is blocking

Issue: calyxos#394
Original-Change-Id: I8194e9b7b64c0a23a42aeb3ffca90602f0882ec4
Change-Id: Ie3452ba56f8ad4786adef004e829d60dee3aa15d
---
 .../systemui/qs/tileimpl/QSTileImpl.java      | 34 ++++++++++++-------
 1 file changed, 21 insertions(+), 13 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
index 1d1e9911884c..13f510017967 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
@@ -32,6 +32,7 @@ import static com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
 import android.annotation.CallSuper;
 import android.annotation.NonNull;
 import android.annotation.SuppressLint;
+import android.app.KeyguardManager;
 import android.content.Context;
 import android.content.Intent;
 import android.graphics.drawable.Drawable;
@@ -39,6 +40,7 @@ import android.metrics.LogMaker;
 import android.os.Handler;
 import android.os.Looper;
 import android.os.Message;
+import android.os.UserHandle;
 import android.text.format.DateUtils;
 import android.util.ArraySet;
 import android.util.Log;
@@ -75,6 +77,8 @@ import com.android.systemui.qs.logging.QSLogger;
 import java.io.PrintWriter;
 import java.util.Objects;
 
+import lineageos.providers.LineageSettings;
+
 /**
  * Base quick-settings tile, extend this to create a new tile.
  *
@@ -288,30 +292,21 @@ public abstract class QSTileImpl<TState extends State> implements QSTile, Lifecy
 
     @Override
     public void click(@Nullable Expandable expandable) {
-        mMetricsLogger.write(populate(new LogMaker(ACTION_QS_CLICK).setType(TYPE_ACTION)
-                .addTaggedData(FIELD_STATUS_BAR_STATE,
-                        mStatusBarStateController.getState())));
-        mUiEventLogger.logWithInstanceId(QSEvent.QS_ACTION_CLICK, 0, getMetricsSpec(),
-                getInstanceId());
         final int eventId = mClickEventId++;
         mQSLogger.logTileClick(mTileSpec, mStatusBarStateController.getState(), mState.state,
                 eventId);
         if (!mFalsingManager.isFalseTap(FalsingManager.LOW_PENALTY)) {
-            mHandler.obtainMessage(H.CLICK, eventId, 0, expandable).sendToTarget();
+            handleClick(ACTION_QS_CLICK, QSEvent.QS_ACTION_CLICK, H.CLICK, eventId, expandable);
         }
     }
 
     @Override
     public void secondaryClick(@Nullable Expandable expandable) {
-        mMetricsLogger.write(populate(new LogMaker(ACTION_QS_SECONDARY_CLICK).setType(TYPE_ACTION)
-                .addTaggedData(FIELD_STATUS_BAR_STATE,
-                        mStatusBarStateController.getState())));
-        mUiEventLogger.logWithInstanceId(QSEvent.QS_ACTION_SECONDARY_CLICK, 0, getMetricsSpec(),
-                getInstanceId());
         final int eventId = mClickEventId++;
         mQSLogger.logTileSecondaryClick(mTileSpec, mStatusBarStateController.getState(),
                 mState.state, eventId);
-        mHandler.obtainMessage(H.SECONDARY_CLICK, eventId, 0, expandable).sendToTarget();
+        handleClick(ACTION_QS_SECONDARY_CLICK, QSEvent.QS_ACTION_SECONDARY_CLICK, H.SECONDARY_CLICK,
+                eventId, expandable);
     }
 
     @Override
@@ -325,7 +320,20 @@ public abstract class QSTileImpl<TState extends State> implements QSTile, Lifecy
         mQSLogger.logTileLongClick(mTileSpec, mStatusBarStateController.getState(), mState.state,
                 eventId);
         if (!mFalsingManager.isFalseLongTap(FalsingManager.LOW_PENALTY)) {
-            mHandler.obtainMessage(H.LONG_CLICK, eventId, 0, expandable).sendToTarget();
+            handleClick(ACTION_QS_LONG_PRESS, QSEvent.QS_ACTION_LONG_PRESS, H.LONG_CLICK,
+                    eventId, expandable);
+        }
+    }
+
+    private void handleClick(int category, QSEvent event, int message, int eventId, Expandable expandable) {
+        final KeyguardManager keyguardManager = mContext.getSystemService(KeyguardManager.class);
+        mMetricsLogger.write(populate(new LogMaker(category).setType(TYPE_ACTION)
+                .addTaggedData(FIELD_STATUS_BAR_STATE, mStatusBarStateController.getState())));
+        mUiEventLogger.logWithInstanceId(event, 0, getMetricsSpec(), getInstanceId());
+        if (!keyguardManager.isKeyguardLocked() ||
+                LineageSettings.Secure.getInt(mContext.getContentResolver(),
+                LineageSettings.Secure.QS_TILES_TOGGLEABLE_ON_LOCK_SCREEN, 1) == 1) {
+            mHandler.obtainMessage(message, eventId, 0, expandable).sendToTarget();
         }
     }
 
-- 
GitLab

